extensions/amp-apester-media/0.1/monetization/ads-constructor.js

```diff
@@ -0,0 +1,246 @@
+      'data-slot': `${slot}`,
+      'data-multi-size-validation': 'false',
+      'data-enable-refresh': `${defaultRefreshTime}`,
+    }
+  );
+  if (rtcConfig) {
+    ampAd.setAttribute('rtc-config', JSON.stringify(rtcConfig));
+  }
+  ampAd.classList.add('i-amphtml-amp-apester-bottom-ad');
+  apesterElement.appendChild(ampAd);
+  Services.mutatorForDoc(apesterElement).requestChangeSize(ampAd, height);
+  return ampAd;
+}
+
+
+/**
+ * @param {!AmpElement} adWrap
+ * @param {!AmpElement} progressBar
+ * @param {!JsonObject} refreshOptions
+ */
+function showInUnitAd(adWrap, progressBar, refreshOptions) {
+  const {skipTimer, timeInView} = refreshOptions;
+  const showTime = timeInView || skipTimer;
+  adWrap.classList.add('active');
+  setStyle(progressBar, 'animation', `progress ${showTime}s linear 1`);
+  const timer = setTimeout(() => {
+    adWrap.classList.remove('active');
+    clearTimeout(timer);
+  }, showTime * 1000);
+}
+
+/**
+ * @param {string} mediaType
+ * @param {string} aniviewPlayerIdOrAdUnit
+ * @param {!AmpElement} apesterElement
+ * @param {!JsonObject} consentObj
+ * @param {!JsonObject} refreshOptions
+ * @param {string} publisherId
+ * @param {!JsonObject} rtcConfig
+ */
+export function constructInUnitAd(mediaType, aniviewPlayerIdOrAdUnit, apesterElement, consentObj, refreshOptions, publisherId, rtcConfig) {
+  let size;
+  let ampAvAd;
+  const {skipTimer, timeout, timeInView, timeBetweenAds} = refreshOptions;
+  const refreshTime = ((timeBetweenAds || timeout || 20) + (timeInView || skipTimer)) * 1000;
+  if (mediaType === AD_TYPES.video) {
+    size = getCompanionVideoAdSize(apesterElement);
+    ampAvAd = getAniviewAdElement(apesterElement, size, aniviewPlayerIdOrAdUnit, consentObj, publisherId);
+  } else {
+    size = {width: 300, height: 250};
+    ampAvAd = createElementWithAttributes(
+      /** @type {!Document} */ (apesterElement.ownerDocument),
+      'amp-ad',
+      {
+        'width': `${size.width}`,
+        'height': `${size.height}`,
+        'type': 'doubleclick',
+        'layout': 'fixed',
+        'data-slot': `${aniviewPlayerIdOrAdUnit}`,
+        'data-multi-size-validation': 'false',
+        'data-enable-refresh': `${refreshTime}`,
+      }
+    );
+    if (rtcConfig) {
+      ampAd.setAttribute('rtc-config', JSON.stringify(rtcConfig));
+    }
+  }
+
+  ampAvAd.classList.add('i-amphtml-amp-apester-in-unit');
+
+  const ampAvAdWrap = createElementWithAttributes(
+    /** @type {!Document} */ (apesterElement.ownerDocument),
+    'div',
+    {'class': 'i-amphtml-amp-apester-in-unit-wrap'}
+  );
+
+  const progressBarWrap = createElementWithAttributes(
+    /** @type {!Document} */ (apesterElement.ownerDocument),
+    'div',
+    {'class': 'i-amphtml-amp-apester-progress-bar'}
+  );
+  ampAvAdWrap.appendChild(progressBarWrap);
+  ampAvAdWrap.appendChild(ampAvAd);
+  apesterElement.appendChild(ampAvAdWrap);
+
+  showInUnitAd(ampAvAdWrap, progressBarWrap, refreshOptions);
+  setInterval(
```

Comment:
@powerivq do you think this should be in a vsync?

Reply:
Yes, that would be better. I do not think we do setInterval in the code base.

Reply:
@powerivq @erwinmombay  
Hey, guys. I don't understand how in this case launch again function after timeout. Could u help me, also it was before I just moved to one utils file

Reply:
@andreyYatskov could you explain why you need interval here? I don't fully understand why `showInUnitAd` needs to be repeatedly called

Reply:
so, here we are show an ad on the same place where unit and cover it, so this is making by some interval apprx each 30sec or 1 min

Reply:
@andreyYatskov Do you mean it shows video ads in the middle of playing? My concerns are:
1. There is no management of this callback wheresoever. Even if the player is removed, the timeout is still there.
2. Repeated polling is not a very efficient use of the resources.

Reply:
@powerivq hey, it was before in another file I just moved to one utils https://github.com/ampproject/amphtml/pull/39814/files#diff-37f33fea038b529eea9fcd8ea1b6a593c41299791d08a2069647f127e55bf253.  Also there no possibility that player will be destroyed.

So basically `showInUnitAd` creating covered background and showing an ad and then hide and should repeat this periodically

Reply:
@andreyYatskov AMP runtime does resource recycling and will destroy things in order to reuse the runtime. Therefore, components should clean up the states when they are being destroyed. Here as well.

Reply:
ok, @powerivq so what u propose to apply? cause I'm not so familiar with amphtml and don't know the best practice and don't have an idea how to implement periodically appearing/hiding an element

Reply:
@andreyYatskov setInterval is okay. You just need to remove the interval by calling `clearTimeout` in your `unlayoutCallback`.

Reply:
agree here, as long as we do cleanup i'd be ok with the change